### ISSUE TITLE
WISH: add string! in draw dialect and simply ignore them

### ISSUE THREAD

**Issue by luce80** [red/red#4661]
I wish `strings!`s be added to `draw` dialect and simply be ignored when parsing them. This way they can act as comments and also as placeholders and could be placed everywhere. This is already implemented in Rebol2 and let you do something like:
```
draw [pen red line 0x0 "end" 0x0]
...
face/effect/draw/("end"): event/offset
```
Note that `get-words!` in current draw implementation do not behave like this.

**Comment by dockimbel** [comment]
Strings are not the right types to be used as markers. Draw dialect already allows set-words to be used for that:
```
draw [pen red line 0x0 end: 0x0]
...
face/effect/draw/end: event/offset
```

**Comment by luce80** [comment]
It is true that strings are not the right types to be used as markers, but they are unharmful, they could be used as "permanent" comments that do not get erased by a `load mold` and IMHO adding them is simple, useful and backward compatible.

**Comment by dockimbel** [comment]
Strings are heavy to store and process in memory compared to words or real comments. The concept of "permanent comment" is vague, and nobody ever seemed to need that so far. If you want to introduce such concept into the language or one of the dialect, you need to write down a sound rationale for it. I see no good reason to support them in this case.

@greggirwin What do you think about that?

**Comment by endo64** [comment]
@dockimbel `set-word!`s _in between_ doesn't seem to work:

```
i: make image! 100x100
;these two works
draw i [pen red line 0x0 0x0 here:]
draw i [pen red here: line 0x0 0x0]

;these two don't work
draw i [pen red line 0x0 here: 0x0]
draw i [pen red line here: 0x0 0x0]
```

I'm not sure if it was working before?

**Comment by 9214** [comment]
@endo64 in-between Draw commands, not their elements.

**Comment by greggirwin** [comment]
I agree with @dockimbel that strings are not the right choice here. If you have large draw blocks to comment, comments are probably a good choice. Strings already have a mean, as data for text commands. If you use them for comments, then you have to be aware of what command is in effect, and adjust their position. e.g. you can't use them as a leading comment for large text.
```
view [
    base draw [
        text 10x10 "big text coming" {hello! and welcome...
        to the new draw example
        }
    ]
]
```

**Comment by 9214** [comment]
I think @luce80's original wish is more concerned with ability to access sections of Draw block on a fine-grained level and to name the elements. IMO `set-word!`s serve this purpose just fine, albeit they can only be placed between commands, which should be relaxed.

The question is: how Draw should process them? Should it simply ignore them, or should it set them to the remainder of the series? In the latter case, should the words itself be removed from a Draw block, or should they stay?

**Comment by greggirwin** [comment]
Use cases will help. I know I've used markers quite a bit at times, but you run into issues of draw blocks that are "chunks" of repeating sections, and what do you do about references there, where a single name makes sense for one chunk, but needs to numbered (or processed by sub-block) in an entire draw block. And how much of this do we put into `draw` versus how much we build at the mezz level with pre-processing.

**Comment by 9214** [comment]
One idea is to make naming relative rather than absolute, by aliasing integer indices with words. Roughly speaking:
```red
set [center radius][2 3]

circle: [circle 50x50 100]
circle/:radius: 30
circle/:center: 10x10
```
IIRC @dockimbel had some ideas in this regard, but I am a bit fuzzy on details at this point.

**Comment by greggirwin** [comment]
I did something like that, at the mezz level, in a couple things. Resize-image and Bubbles.

**Comment by luce80** [comment]
My use case in R2 is http://www.rebol.org/view-script.r?script=color-requester.r . But even without this use case, IMHO adding them is simple (simply ignore them (apart for text !!)), unharmful, useful and **backward compatible**. Any other similar enhancement is also welcome of course, I am not fixed on string!s.

**Comment by greggirwin** [comment]
Thanks @luce80. That's helpful.

If we want more general markers, we can either relax the rules as @9214 noted, or use another type. The only other word types are issue! and refinement!, and issues are used for hex colors as well, so that leaves refinements. It's a _little_ odd to use them this way. Where it gets ugly is incorporating them into paths. e.g.
```
>> blk: [a b /c [d e /f [g h] ff] cc]
== [a b /c [d e /f [g h] ff] cc]
>> blk/c
== none
>> blk/(/c)
== [d e /f [g h] ff]
>> blk/(/c)/(/f)
== [g h]
; Or with indirection
>> key-c: /c
== /c
>> blk/:key-c/(/f)
== [g h]
```

@dockimbel @9214 @endo64 what are the impacts of relaxing set-word! use, and would a possible middle-ground be to extend their support to sub-blocks that follow them?

**Comment by 9214** [comment]
The way Draw is implemented ATM, `set-word!`s need to be preprocessed and set to series' index, otherwise the core loop needs to be rewritten and include `TYPE_SET_WORD` branch for every case, which is quite costly.

Restricting their usage to sub-blocks would require to keep the context: "if it's a block and it was preceded by `set-word!`", or "if it's a `set-word!` followed by a `block!`".

The main design question IMO is if such markers should be internal (e.g. abovementioned `refinement!`) or external (e.g. integer/symbol aliases) to the Draw block: the latter seems more general and not dialect-specific, but requires a new language feature.

**Comment by greggirwin** [comment]
Thanks @9214. Sounds to heavy to change the use of set-words. As `draw` use grows, and is leveraged by more people, we have to keep performance in mind. e.g. charting stuff Bolek and Harald are working on.

On refinements, is there any other use case we can think of where they might be added to the `draw` dialect as a new feature? If not, the two issues are the ugliness of accessing them and breaking code that uses them as ad-hoc keys if we do fine a use case for them. Symbol aliases also seem too heavy a change for this, and more maintenance going forward. And how do we avoid naming conflicts while keeping symbols precise? It's a bit of worrying about a problem we don't have yet, but if `draw` just ignores refinements, it's more of an "it's all just data" view.
